docs/implementation.adoc

```diff
@@ -0,0 +1,507 @@
+== Implementation Guide
+
+This document provides the guidelines for a correct optimal implementation of the SCM API.
+
+=== Introduction
+
+The SCM API consists of a number of extension points:
+
+* The original `hudson.scm.SCM` extension point is responsible for:
+
+** performing checkouts from source control into a `FilePath`
+
+** calculating the changelogs between revisions
+
+** constructing the changelog parser
+
+* The `jenkins.scm.api.SCMSource` extension point is responsible for:
+
+** identifying heads / branches that are available
+
+** tracking the current revisions of heads
+
+** constructing `hudson.scm.SCM` instances for a given head and revision pair
+
+* The `jenkins.scm.api.SCMNavigator` extension point is responsible for:
+
+** enumerating potential `jenkins.scm.api.SCMSource` instances
+
+Some examples of how these concepts may be mapped into well known source control systems:
+
+* Git has a fairly direct mapping of concepts
+
+** `jenkins.scm.api.SCMHead` would represent the branches and tags in an individual Git repository
+
+** `jenkins.scm.api.SCMRevision` would represent the Git commit hash of each individual commit. Because Git commits are immutable, each `SCMRevision` is deterministic as if you ask to check out the same hash you will always get the same contents.
```

Comment:
For mergeable PRs in `github-branch-source` it is actually a pair of commits—base branch and proposed branch. Worth mentioning that these things can be composite structures.